theano/tensor/nnet/abstract_conv.py

```diff
@@ -2094,15 +2223,44 @@ def perform(self, node, inp, out_):
                               for i in range(self.convdim))] = topgrad
             topgrad = new_topgrad
 
-        axes_order = (1, 0) + tuple(range(2, self.convdim + 2))
-        flip_filters = ((slice(None), slice(None)) +
-                        (slice(None, None, -1),) * self.convdim)
-        kern = kern.transpose(axes_order)
-        if self.filter_flip:
-            topgrad = topgrad[flip_filters]
-        img = self.conv(topgrad, kern, mode="full", dilation=self.filter_dilation)
-        if self.filter_flip:
-            img = img[flip_filters]
+        if self.unshared is True:
+            img = np.zeros(imshp, dtype=topgrad.dtype)
+            topgrad_pad = np.zeros((topgrad.shape[0], topgrad.shape[1],
+                                    topgrad.shape[2] + (kern.shape[4] - 1) * 2,
+                                    topgrad.shape[3] + (kern.shape[5] - 1) * 2),
+                                   dtype=topgrad.dtype)
+            topgrad_pad[:, :, (kern.shape[4] - 1):(1 - kern.shape[4]),
+                        (kern.shape[5] - 1):(1 - kern.shape[5])] = topgrad
+            topgrad = topgrad_pad
+            if not self.filter_flip:
+                kern = kern[(slice(None),) * 4 +
+                            (slice(None, None, -1),) * self.convdim]
+            for b in xrange(topgrad.shape[0]):
+                for n in xrange(topgrad.shape[1]):
+                    for im0 in xrange(kern.shape[1]):
+                        for row in xrange(img.shape[2]):
+                            for col in xrange(img.shape[3]):
+                                for k_row in xrange(kern.shape[4]):
+                                    for k_col in xrange(kern.shape[5]):
+                                        # Deciding which kernel to select
+                                        reg_row = row - ((kern.shape[4] - 1) - k_row)
+                                        reg_col = col - ((kern.shape[5] - 1) - k_col)
+                                        reg_row = max(0, min(reg_row, kern.shape[2] - 1))
+                                        reg_col = max(0, min(reg_col, kern.shape[3] - 1))
+                                        img[b, im0, row, col] += kern[n, im0, reg_row,
+                                                                      reg_col, k_row, k_col] * \
+                                            topgrad[b, n, row + k_row, col + k_col]
+        else:
+            flip_filters = ((slice(None), slice(None)) +
+                            (slice(None, None, -1),) * self.convdim)
+            if self.filter_flip:
+                topgrad = topgrad[flip_filters]
+            axes_order = (1, 0) + tuple(range(2, self.convdim + 2))
+            kern = kern.transpose(axes_order)
+            img = self.conv(topgrad, kern, mode="full", dilation=self.filter_dilation)
+            if self.filter_flip:
+                img = img[flip_filters]
```

Comment:
I have a doubt here. Why do we flip the topgrad, convolve and then again flip the output? Wouldn't it be easier to just flip (or not flip) the kern instead?